services/metadata/src/metadata/persistence/avu.clj

```diff
@@ -98,25 +103,38 @@
       existing-avu
       (add-avus user-id [avu]))))
 
+(defn- add-orphaned-ids-where-clause
+  [query avu-ids-to-keep]
+  (if (empty? avu-ids-to-keep)
+    query
+    (where query {:id [not-in avu-ids-to-keep]})))
+
 (defn remove-orphaned-avus
-  "Removes AVUs for the given target-id that are not in the given set of avu-ids."
-  [target-type target-id avu-ids]
-  (let [query (-> (delete* :avus)
-                  (where {:target_id target-id
-                          :target_type (db/->enum-val target-type)}))]
-    (if (empty? avu-ids)
-      (delete query)
-      (delete (where query {:id [not-in avu-ids]})))))
+  "Removes AVUs for the given target-id that are not in the given set of avu-ids-to-keep."
+  [target-type target-id avu-ids-to-keep]
+  (let [avus-to-remove (-> (select* :avus)
+                           (target-where-clause target-type target-id)
+                           (add-orphaned-ids-where-clause avu-ids-to-keep)
+                           select)]
+    ;; Remove orphaned sub-AVUs of any AVUs to be removed by this request.
+    (doseq [{:keys [id]} avus-to-remove]
+      (remove-orphaned-avus "avu" id nil))
```

Comment:
I don't know what korma provides for us, but in general I'd implement this with a recursive CTE passed to the delete, letting postgres do the work. Roughly:

```sql
WITH RECURSIVE orphaned AS
  (SELECT id FROM avus WHERE target_id = $1 AND target_type = $2 AND <avus to keep etc>
     UNION
   SELECT id FROM avus WHERE target_type = "avu" AND target_id IN (SELECT id FROM orphaned))
DELETE FROM avus WHERE id IN (SELECT id FROM orphaned)
```

but, I don't know if this is worth the time implementing/testing